pkg/image/registry/imagestream/strategy.go

```diff
@@ -229,19 +230,31 @@ func tagReferenceToTagEvent(stream *api.ImageStream, tagRef api.TagReference, ta
 		}, nil
 
 	case "ImageStreamImage":
-		ref, err := api.DockerImageReferenceForStream(stream)
-		if err != nil {
-			return nil, err
+		var event *api.TagEvent
+		// while searching for all the image IDs in stream that start with imageID
+		// save the tag event for proper tagging, code borrowed from api.ResolveImageID
+		resolvedIDs := sets.NewString()
+		for _, history := range stream.Status.Tags {
+			for _, tagging := range history.Items {
+				if d, err := digest.ParseDigest(tagging.Image); err == nil {
+					if strings.HasPrefix(d.Hex(), tagOrID) || strings.HasPrefix(tagging.Image, tagOrID) {
+						event = &tagging
+						resolvedIDs.Insert(tagging.Image)
+					}
+					continue
+				}
+				if strings.HasPrefix(tagging.Image, tagOrID) {
+					event = &tagging
+					resolvedIDs.Insert(tagging.Image)
+				}
+			}
 		}
-
-		resolvedIDs := api.ResolveImageID(stream, tagOrID)
```

Comment:
Actually both of these places expect single result, others (0 or 2+) are interpreted as errors. If then ResolveImageID would actually return the tagevent rather than just the ID, it does now. Additionally returning error, but since rest.go needs distinction between 0 and 2+, we could return NotFound and Conflict errors appropriately, wdyt?